From 97917c57305e7b85f35570c84507fc9f3a19ea7d Mon Sep 17 00:00:00 2001
From: Sivasubramanaiyan Padmanaban <sivasubramaniyan.padmanaban@analog.com>
Date: Fri, 14 Aug 2026 11:48:38 +0530
Subject: [PATCH] drivers: media: i2c: nv_adsd3500: enable streaming before
 starting FSYNC PWM

- Send the STREAM-ON command before enabling the FSYNC PWM output.
- One frame-duration delay to allow the sensor to enter streaming mode.
- Ensure PWM-triggered synchronization starts only after streaming is active.

Signed-off-by: Sivasubramanaiyan Padmanaban <sivasubramaniyan.padmanaban@analog.com>
---
 drivers/media/i2c/adsd3500.c | 13 +++++++------
 1 file changed, 7 insertions(+), 6 deletions(-)

diff --git a/drivers/media/i2c/adsd3500.c b/drivers/media/i2c/adsd3500.c
index f54edada3b1d..34435886881d 100755
--- a/drivers/media/i2c/adsd3500.c
+++ b/drivers/media/i2c/adsd3500.c
@@ -1556,7 +1556,14 @@ static int adsd3500_start_streaming(struct adsd3500 *adsd3500)
 
 	dev_info(adsd3500->dev, "%s\n", __func__);
 
+	ret = regmap_write(adsd3500->regmap, STREAM_ON_CMD, STREAM_ON_VAL);
+	if (ret < 0) {
+		dev_err(adsd3500->dev, "Write of STREAM-ON command failed.\n");
+		return ret;
+	}
+
 	if(adsd3500->pwm_fsync != NULL && adsd3500->curr_sync_mode == PWM_TRIGGER ){
+		msleep(5);
 		ret = pwm_enable(adsd3500->pwm_fsync);
 		if (ret) {
 			dev_err(adsd3500->dev, "Could not enable FSYNC PWM\n");
@@ -1564,12 +1571,6 @@ static int adsd3500_start_streaming(struct adsd3500 *adsd3500)
 		}
 	}
 
-	ret = regmap_write(adsd3500->regmap, STREAM_ON_CMD, STREAM_ON_VAL);
-	if (ret < 0) {
-		dev_err(adsd3500->dev, "Write of STREAM-ON command failed.\n");
-		return ret;
-	}
-
 	return ret;
 }
 
-- 
2.25.1

